vowpalwabbit/io/io_adapter.cc

```diff
@@ -0,0 +1,415 @@
+#include "io_adapter.h"
+
+#ifdef _WIN32
+#define NOMINMAX
+#define ssize_t int64_t
+#include <winsock2.h>
+#include <io.h>
+#else
+#include <sys/socket.h>
+#include <unistd.h>
+#endif
+
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <fcntl.h>
+
+#include <cstdio>
+#include <cassert>
+#include <cstring>
+#include <iostream>
+#include <fstream>
+#include <vector>
+#include <algorithm>
+
+#include <zlib.h>
+#if (ZLIB_VERNUM < 0x1252)
+using void* gzFile;
+#else
+struct gzFile_s;
+typedef struct gzFile_s* gzFile;
+#endif
+
+#ifndef O_LARGEFILE  // for OSX
+#define O_LARGEFILE 0
+#endif
+
+using namespace VW::io;
+
+enum class file_mode
+{
+  read,
+  write
+};
+
+struct socket_adapter : public writer, public reader
+{
+  socket_adapter(int fd, const std::shared_ptr<details::socket_closer>& closer)
+      : reader(false /*is_resettable*/), _socket_fd{fd}, _closer{closer}
+  {
+  }
+  ssize_t read(char* buffer, size_t num_bytes) override;
+  ssize_t write(const char* buffer, size_t num_bytes) override;
+
+private:
+  int _socket_fd;
+  std::shared_ptr<details::socket_closer> _closer;
+};
+
+struct stdio_adapter : public writer, public reader
+{
+  stdio_adapter() : reader(false /*is_resettable*/) {}
+  ssize_t read(char* buffer, size_t num_bytes) override;
+  ssize_t write(const char* buffer, size_t num_bytes) override;
+};
+struct file_adapter : public writer, public reader
+{
+  // investigate whether not using the old flags affects perf. Old claim:
+  // _O_SEQUENTIAL hints to OS that we'll be reading sequentially, so cache aggressively.
+  file_adapter(const char* filename, file_mode mode);
+  file_adapter(int file_descriptor, file_mode mode);
+  ~file_adapter();
+  ssize_t read(char* buffer, size_t num_bytes) override;
+  ssize_t write(const char* buffer, size_t num_bytes) override;
+  void reset() override;
+
+private:
+  int _file_descriptor;
+  file_mode _mode;
+};
+
+struct gzip_file_adapter : public writer, public reader
+{
+  gzip_file_adapter(const char* filename, file_mode mode);
+  gzip_file_adapter(int file_descriptor, file_mode mode);
+  ~gzip_file_adapter();
+
+  ssize_t read(char* buffer, size_t num_bytes) override;
+  ssize_t write(const char* buffer, size_t num_bytes) override;
+  void reset() override;
+
+private:
+  gzFile _gz_file;
+  file_mode _mode;
+};
+
+struct gzip_stdio_adapter : public writer, public reader
+{
+  gzip_stdio_adapter();
+  ~gzip_stdio_adapter();
+  ssize_t read(char* buffer, size_t num_bytes) override;
+  ssize_t write(const char* buffer, size_t num_bytes) override;
+
+private:
+  gzFile _gz_stdin;
+  gzFile _gz_stdout;
+};
+
+struct vector_writer : public writer
+{
+  vector_writer(std::vector<char>& buffer);
+  ~vector_writer() = default;
+  ssize_t write(const char* buffer, size_t num_bytes) override;
+
+private:
+  std::vector<char>& _buffer;
```

Comment:
It is dangerous, but I can't think of another way to let the caller have access to the buffer written into. Ideas?

Reply:
If you don't want to leak implementation details, I can see 2 options here.
1. Make this a shared_ptr
2. Have the caller implement their own writer that will give them access to the buffer. Not ideal, but a lot better than providing an inherently dangerous utility out of the box

Reply:
A `shared_ptr` seems pretty reasonable here I think

Reply:
I think `shared_ptr` is a good compromise.

Another possibility (better encapsulation, but more restrictive and could have undefined behavior) would be to add a `const char* get_buffer()` function to the interface. It would have to be noted that the pointer is invalidated if anything modifies the object though

Reply:
It would be rather difficult to make `const char* get_buffer()` make sense for all writers